lxd/ovn.go

```diff
@@ -0,0 +1,157 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/canonical/lxd/lxd/state"
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/api"
+	"github.com/canonical/lxd/shared/logger"
+)
+
+type OVNLogMonitor struct {
+	state         *state.State
+	paths         []string
+	initialStats  map[string]fs.FileInfo
+	lastTimestamp time.Time
+}
+
+// NewOVNLogMonitor returns a new OVNLogMonitor.
+func NewOVNLogMonitor(s *state.State) *OVNLogMonitor {
+	// Use the logs from MicroOVN if available, otherwise use the ones in /var/log/ovn.
+	ovnLogPath := "/var/snap/microovn/logs"
+
+	if !shared.PathExists(ovnLogPath) {
+		ovnLogPath = "/var/log/ovn"
+	}
+
+	return &OVNLogMonitor{
+		state: s,
+		paths: []string{
+			filepath.Join(ovnLogPath, "ovn-controller.log"),
+			filepath.Join(ovnLogPath, "ovn-northd.log"),
+			filepath.Join(ovnLogPath, "ovsdb-server-nb.log"),
+			filepath.Join(ovnLogPath, "ovsdb-server-sb.log"),
+		},
+		initialStats:  make(map[string]fs.FileInfo),
+		lastTimestamp: time.Now(),
+	}
+}
+
+// Start starts the log file monitor. It checks whether some file stats have changed, and sends the log entries to the events API.
+func (l *OVNLogMonitor) Start(ctx context.Context) context.CancelFunc {
+	ctx, cancelFunc := context.WithCancel(ctx)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				// fallthrough
+			}
+
+			for _, path := range l.paths {
+				if !l.fileHasChanged(path) {
```

Comment:
Can we use fsmonitor to watch if the files have changed rather than polling and sleeping?